### PR TITLE
API for Simulation

### LINKED ISSUE #13: Feature: Implement the endpoints for the simulator
**Describe the solution you'd like**
All endpoints should be implemented as described in: [this file](https://github.com/itu-devops/lecture_notes/blob/master/sessions/session_03/README_TASKS.md)

**Additional context**
This is done when issues #8 #10 #11 #12 are completed

### PR DESCRIPTION
**Pull request by nikso-itu**
# Description

Implements all endpoints that are required for the simulation.

Closes #13, implements #24, implements #18, implements #15, implements #16, implements #17 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by bitknox — APPROVED** [review]
All good. Code has been tested with the test_sim_compliance.py script.

**Review by Mansin-ITU — APPROVED** [review]
Looks good to me!

**Review by KjellbergD — APPROVED** [review]

File: controllers/messageController.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+		}
+
+		// get DB
+		db := helpers.GetTypedDb(c)
+
+		// get username, and convert to user id
+		username := c.Param("username")
+		userId, err := helpers.GetUserId(db, username)
+		if err != nil {
+			// abort if unsuccessful
```
Comment (Lindharden) [inline controllers/messageController.go:114]:
```suggestion
```

File: controllers/messageController.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+			c.AbortWithStatus(404)
+		}
+
+		// get current time as Unix time
```
Comment (Lindharden) [inline controllers/messageController.go:118]:
```suggestion
```

File: controllers/messageController.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+		time := time.Now().Unix()
+
+		// insert into DB
```
Comment (Lindharden) [inline controllers/messageController.go:121]:
```suggestion
```

File: helpers/authenticationHelpers.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+	if !checkUserPasswords(password, password2) {
+		return model.User{}, errors.New("the two passwords do not match")
+	}
+
+	if !checkUserEmail(email) {
+		return model.User{}, errors.New("you have to enter a valid email address")
+	}
+
+	if checkUsernameExists(db, username) {
```
Comment (Lindharden) [inline helpers/authenticationHelpers.go:26]: The order of these checks (username, email & password) should most likely match the order of the inputs in the register form. The user shouldn’t be told about his wrongly typed re-password before being told that his username is already taken.

**Review by Lindharden — APPROVED** [review]
Looks good

**Comment by Lindharden** [comment]
We should probably also extract the simulator methods to a controller only responsible for simulation to ensure cohesion

**Pull request merged**